stan/math/prim/meta/apply_scalar_binary.hpp

```diff
@@ -51,6 +51,12 @@ inline auto apply_scalar_binary(const T1& x, const T2& y, const F& f) {
 template <typename T1, typename T2, typename F,
           require_all_eigen_t<T1, T2>* = nullptr>
 inline auto apply_scalar_binary(const T1& x, const T2& y, const F& f) {
+  if ((x.rows() != y.rows()) || (x.cols() != y.cols())) {
+    std::ostringstream msg;
+    msg << "Inputs to vectorized binary function must match in"
+        << " size if one is not a scalar";
+    throw std::domain_error(msg.str());
+  }
```

Comment:
Don't we have some check function to turn this into a single line?

Reply:
We do, but all the check functions include ```prim/meta.hpp```, which includes this header, and spirals into circular dependencies

Reply:
Hmm, do we need this to be in meta? I think it would belong into functor.

Reply:
I had it in meta to be consistent with ```apply_scalar_unary```. So if I move this one I'd have to move the other two, and I'm a bit hesitant to be moving existing headers after all of the issues that moving the ```Eigen.hpp``` headers has caused with RStan

Reply:
I think apply_scalar_unary should also eventually be moved (no need to do it in this PR). I think reducing code duplication should be a priority compared to being consistent with what turns out to be a poor design choice.

I am not saying you must change this, but I'd prefere if you did.

Reply:
Works for me, I can move those files in this pull

Reply:
I've moved the ```apply_*``` headers to the functor folder and updated the other functions that use them. This is ready for another review